File: Data Strucure/03.Linear Data Structures - Stacks and Queues/CSharp/Stack/Stack/Stack.cs
using System;
using System.Collections;
using System.Collections.Generic;
public class StaticStack<T> : IEnumerable<T>
{
    private int MIN_CAPACITY = 4;
    private T[] array;
    private int index;

    public int Count { get; set; }
    public int Capacity => this.array.Length;

    public StaticStack()
    {
        this.array = new T[MIN_CAPACITY];
        index = 0;
    }
    public void Push(T item)
    {
        if (this.Count >= this.array.Length)
        {
            Resize();
        }
        this.array[index++] = item;

        this.Count++;
    }

    private void Resize()
    {
        T[] newArray = new T[this.array.Length * 2];
        for (int i = 0; i < this.array.Length; i++)
        {
            newArray[i] = this.array[i];
        }
        this.array = newArray;
    }

    public T Pop()
    {
        if (this.Count == 0 || index < 0)
        {
            throw new InvalidOperationException("Empty stack");
        }
        T element = this.array[index - 1];
        this.array[index--] = default(T);
        this.Count--;
        return element;
    }

    public T Peek()
    {
        if (this.Count == 0 || index < 0)
        {
            throw new InvalidOperationException("Empty stack");
        }
        T element = this.array[index-1];

        return element;
    }

    public void TrimExcess()
    {
        T[] newArray = new T[this.Count];
        int last = newArray.Length - 1;

        for (int i = this.array.Length - 1; i > index; i--)
        {
            newArray[last--] = this.array[i];
        }
        this.array = newArray;
    }

    public IEnumerator<T> GetEnumerator()
    {
        for (int i = 0; i < this.array.Length; i++)
        {
            yield return this.array[i];
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }
}

File: Data Structures Fundamentals (with Java)/03.Linear-Data-Structures/02. Skeleton/src/main/java/implementations/Stack.java
package implementations;
import interfaces.AbstractStack;
import java.util.Iterator;

public class Stack<E> implements AbstractStack<E> {
    private E[] data;
    private int size;
    private int index;
    private int capacity = 4;

    public Stack() {
        this.data =  (E[]) new Object[this.capacity];
    }

    @Override
    public void push(E element) {
        if (this.size == this.capacity) {
            this.data = resize();
            this.capacity *= 2;
        }

        this.data[this.size] = element;
        this.size++;
    }

    @Override
    public E pop() {
        if (this.isEmpty()) {
            throw new IllegalStateException("Empty stack");
        }
        var top = this.data[this.size - 1];
        this.data[this.size - 1] = null;
        this.size--;
        return top;
    }

    @Override
    public E peek() {
        if (this.isEmpty()) {
            throw new IllegalStateException("Empty stack");
        }
        return this.data[this.size - 1];
     }

    @Override
    public int size() {
        return this.size;
    }

    @Override
    public boolean isEmpty() {
        return this.size() == 0;
    }

    @Override
    public Iterator<E> iterator() {
        return new ListIterator();
    }

    private E[] resize() {
        var newArray = (E[]) new Object[this.size * 2];
        for (int i = 0; i < this.data.length; i++) {
            newArray[i] = this.data[i];
        }
        return newArray;
    }

    private final class ListIterator implements Iterator<E> {
        @Override
        public boolean hasNext() {
            return size() > 0;
        }

        @Override
        public E next() {
            return pop();
        }
    }
}
